peps/pep-0667.rst

```diff
@@ -201,15 +413,222 @@ will continue to work correctly. Debuggers and other tools that use
+``globals()`` and ``locals()`` in the calling Python frame as their default
+execution namespace.
+
+This was historically also equivalent to using the calling frame's
+``frame.f_globals`` and ``frame.f_locals`` attributes, but this PEP maps
+them to ``globals()`` and ``locals()`` in order to preserve the property
+of ignoring attempted writes to the local namespace by default.
+
+However, as noted above for ``locals()``, this change has an additional
+effect: each ``exec()`` call in an optimized scope will now run in a
+*different* implicit namespace rather than a shared one. Furthermore,
+separately calling ``locals()`` will also return a different namespace.
+
+This poses a potential compatibility issue for some code, as with the
+current implementation returning the same dict when ``locals()`` is called
+multiple times in function scope, the following code usually works due to
+the implicitly shared local variable namespace::
+
+    def f():
+        exec('a = 0')  # equivalent to exec('a = 0', globals(), locals())
+        exec('print(a)')  # equivalent to exec('print(a)', globals(), locals())
+        print(locals())  # {'a': 0}
+        # However, print(a) will not work here
+    f()
+
+With ``locals()`` in an optimised scope returning the same shared dict for each call,
+it is possible to store extra "fake locals" in that dict. While these aren't real
+locals known by the compiler (so they can't be printed with code like ``print(a)``),
+they can still be accessed via ``locals()`` and shared between multiple ``exec()``
+calls in the same function scope. Furthermore, because they're *not* real locals,
+they don't get implicitly updated or removed when the shared cache is refreshed
+from the local variable storage array.
+
+When the code in ``exec()`` tries to write to an existing local variable, the
+runtime behaviour gets harder to predict::
+
+    def f():
+        a = None
+        exec('a = 0')  # equivalent to exec('a = 0', globals(), locals())
+        exec('print(a)')  # equivalent to exec('print(a)', globals(), locals())
+        print(locals())  # {'a': None}
+    f()
+
+``print(a)`` will print ``None`` because the implicit ``locals()`` call in
+``exec()`` refreshes the cached dict with the actual values on the frame.
+This means that, unlike the "fake" locals created by writing back to ``locals()``
+(including via previous calls to ``exec()``), the real locals known by the
+compiler can't easily be modified by ``exec()``.
+
+As noted in the Motivation section, this confusing side effect happens even if the
+local variable is only defined *after* the ``exec`` calls::
+
+    >>> def f():
+    ...     exec("a = 0")
+    ...     exec("print('a' in locals())") # Printing 'a' directly won't work
+    ...     print(locals())
+    ...     a = None
+    ...     print(locals())
+    ...
+    >>> f()
+    False
+    {}
+    {'a': None}
+
+Because ``a`` is a real local variable, it gets removed from ``locals()`` when
```

Comment:
I had to read this paragraph a few times for it to make sense.  I can't think of a better way to rephrase it, but just wanted to point out my confusion on a first read.

Reply:
I had a go at rewording it (including pointing out that the behaviour is necessary to make `locals()` correctly reflect `del` statements in optimized scopes)

Reply:
Thanks, I think the rewording helps!